File: sandbox/algoritms/bfs.js
const isSeller = (person) => {
    // Function checking if a given person is a seller (condition to meet)
    return person.endsWith("m"); // Example condition - can be adjusted
  };
  
  const search = (name, graph) => {
    // We create a queue to store people to check
    const searchQueue = [...graph[name]];
    
    // We create a list of searched people to avoid cycles
    const searched = [];
  
    while (searchQueue.length > 0) {
      const person = searchQueue.shift(); // We get the first person from the queue
      
      // We check if the person has not been searched yet
      if (!searched.includes(person)) {
        // We check if the person is a seller
        if (isSeller(person)) {
          console.log(`${person} sells mangoes!`);
          return true;
        } else {
          // We add the person's friends to the queue
          searchQueue.push(...graph[person]);
          // We mark this person as searched
          searched.push(person);
        }
      }
    }
  
    return false; // Returns false if no seller is found
  };
  
  // Example graph
  const graph = {
    "ty": ["alice", "bob", "claire"],
    "alice": ["peggy"],
    "bob": ["anuj", "peggy"],
    "claire": ["thom", "jonny"],
    "anuj": [],
    "peggy": [],
    "thom": [],
    "jonny": []
  };
  
  // Running the search
  search("ty", graph);
  

File: Projects/Testtrack/miniS - Algoritms sandbox/bfs.py
from collections import deque


def _person_is_seller(name: str) -> bool:
    """
    Checks is person is a seller. In this example, anyone whom name ends on "m" letter is a seller
    """
    return name[-1] == 'm'


def bfs(start: str, graph: dict) -> str:
    """
    Performs a breadth-first search on the given graph starting from the specified start node.

    Args:
        start (str): The key of the start node of the search.
        graph (dict): A dictionary representing the graph, where each key is a node and each value is a list of neighboring nodes.

    Returns:
        str: A string indicating whether a seller was found or not.
    """
    search_queue = deque()
    search_queue += graph[start]

    while search_queue:
        person = search_queue.popleft()
        if _person_is_seller(person):
            return f"Person {person} is a seller"
        else:
            search_queue += graph[person]

    return "No seller found here"


if __name__ == "__main__":
    graph = {
        "ty": ["alice", "bob", "claire"],
        "alice": ["peggy"],
        "bob": ["anuj", "peggy"],
        "claire": ["thom", "jonny"],
        "anuj": [],
        "peggy": [],
        "thom": [],
        "jonny": []
    }

    print(bfs("ty", graph))
